components/script/dom/bindings/serializable.rs

```diff
@@ -5,27 +5,52 @@
 //! Trait representing the concept of [serializable objects]
 //! (<https://html.spec.whatwg.org/multipage/#serializable-objects>).
 
+use js::jsapi::JSStructuredCloneReader;
+
 use crate::dom::bindings::reflector::DomObject;
-use crate::dom::bindings::structuredclone::StructuredDataHolder;
+use crate::dom::bindings::root::DomRoot;
+use crate::dom::bindings::structuredclone::{
+    CloneableObject, StructuredReadDataHolder, StructuredWriteDataHolder,
+};
 use crate::dom::globalscope::GlobalScope;
 
-/// The key corresponding to the storage location
-/// of a serialized platform object stored in a StructuredDataHolder.
-#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
-pub struct StorageKey {
-    pub index: u32,
-    pub name_space: u32,
+/// An interface to construct a stream of serialization operations that will be
+/// evaluated as part of serializing a DOM object.
+pub trait ToSerializeOperations {
+    fn to_serialize_operations(&self) -> Vec<SerializeOperation>;
+}
+
+/// An interface to reconstruct a known type using previously-serialized values
+/// that can be obtained from the provided JSStructuredCloneReader.
+///
+/// Safety:
+/// The provided pointer must point to a live JSStructuredCloneReader object.
+pub trait FromStructuredClone: Sized {
+    unsafe fn from_structured_clone(r: *mut JSStructuredCloneReader) -> Self;
+}
+
+/// Operations permitted as part of structured clones to serialize arbitrary values
+/// during serialization of DOM objects.
+pub enum SerializeOperation {
+    Uint32Pair(u32, u32),
+    Double(f64),
 }
```

Comment:
This seems like a nice abstraction. The one concern I have is that it means that all data will have to be copied during serialization (apart from blobs, right?). Is this serialization performance sensitive. If it is, another alternative would be to expose matching APIs on the writer and reader like this:

```
impl Writer {
   fn write_uint32_pair(&self, a: u32, b:u32);
}

impl Reader {
   fn read_uint32_pair(&self) -> Result<(u32, u32), Err>
}
```

Then all unsafe code could be confined to the reader and the writer. Maybe there's a reason why this isn't feasible though.